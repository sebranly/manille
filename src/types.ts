export enum CardSuit {
  Clubs = 'clubs',
  Diamonds = 'diamonds',
  Hearts = 'hearts',
  Spades = 'spades'
}

export enum CardRank {
  Ace = 'ace',
  Eight = 'eight',
  Jack = 'jack',
  King = 'king',
  Nine = 'nine',
  Queen = 'queen',
  Seven = 'seven',
  Ten = 'ten'
}

export interface Card {
  rank: CardRank;
  suit: CardSuit;
}

export interface Player {
  id: number;
  cards: Card[];
  name: string;
}

export interface Turn {
  tbd: any;
}

export interface Round {
  turns: Turn[];
}

export interface Game {
  players: Player[];
  rounds: Round[];
}

export interface InfoSuitHighest {
  clubs?: CardRank;
  diamonds?: CardRank;
  hearts?: CardRank;
  spades?: CardRank;
}
