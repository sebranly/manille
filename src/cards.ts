import { Card, CardRank, CardSuit } from './types';
import { CARDS_PER_DECK } from './constants';
import { compareCardRanks } from './scores';
import { flattenArray } from './utils';

export const generateSuit = (suit: CardSuit) => {
  const { Ace, Eight, Jack, King, Nine, Queen, Seven, Ten } = CardRank;

  const cards: Card[] = [
    { rank: Ten, suit },
    { rank: Ace, suit },
    { rank: King, suit },
    { rank: Queen, suit },
    { rank: Jack, suit },
    { rank: Nine, suit },
    { rank: Eight, suit },
    { rank: Seven, suit }
  ];

  return cards;
};

export const generateDeck = () => {
  const { Clubs, Diamonds, Hearts, Spades } = CardSuit;

  const cards: Card[] = flattenArray([Clubs, Diamonds, Spades, Hearts].map(generateSuit));

  return cards;
};

export const cutDeck = (cards: Card[], cardsBefore: number) => {
  const isInvalidCardsBefore = cardsBefore <= 0 || cardsBefore >= CARDS_PER_DECK;
  const safeCardsBefore = isInvalidCardsBefore ? CARDS_PER_DECK / 2 : cardsBefore;

  const part1 = cards.slice(0, safeCardsBefore);
  const part2 = cards.slice(safeCardsBefore);

  return [...part2, ...part1];
};

export const filterBySuit = (cards: Card[], suit: CardSuit | false) => {
  const cardsSuit = cards.filter((card: Card) => card.suit === suit);

  return cardsSuit;
};

export const excludeSuit = (cards: Card[], suit: CardSuit) => {
  const cardsOtherSuits = cards.filter((card: Card) => card.suit !== suit);

  return cardsOtherSuits;
};

export const excludeSuitOver = (cards: Card[], suit: CardSuit, cardRank: CardRank) => {
  const cardsOtherSuitsOrSameSuitLow = cards.filter((card: Card) => {
    return card.suit !== suit || compareCardRanks(cardRank, card.rank) > -1;
  });

  return cardsOtherSuitsOrSameSuitLow;
};

export const excludeCards = (cards: Card[], cardsToExclude: Card[]) => {
  const newCards = cards.filter((card: Card) => !hasCard(cardsToExclude, card));

  return newCards;
};

export const getCardIndex = (cards: Card[], card: Card) => {
  const cardId = cards.findIndex((oneCard: Card) => isSameCard(card, oneCard));

  return cardId;
};

export const isSameCard = (card1: Card, card2: Card) => {
  return card1.rank === card2.rank && card1.suit === card2.suit;
};

export const areSameOrderedCards = (cards1: Card[], cards2: Card[]) => {
  if (cards1.length !== cards2.length) return false;
  return cards1.every((c, i) => isSameCard(c, cards2[i]));
};

export const hasCard = (cards: Card[], card: Card) => {
  const hasIt = cards.some((oneCard: Card) => isSameCard(card, oneCard));

  return hasIt;
};

export const orderCards = (cards: Card[]) => {
  const { Clubs, Diamonds, Hearts, Spades } = CardSuit;

  const clubs = filterBySuit(cards, Clubs);
  const diamonds = filterBySuit(cards, Diamonds);
  const hearts = filterBySuit(cards, Hearts);
  const spades = filterBySuit(cards, Spades);

  const newCardsGroups = [clubs, diamonds, spades, hearts];

  const orderedCardsGroup = newCardsGroups.map(sortSuit);

  return flattenArray(orderedCardsGroup);
};

export const sortSuit = (cardsSuit: Card[]) => {
  return cardsSuit.sort((a, b) => -1 * compareCardRanks(a.rank, b.rank));
};
