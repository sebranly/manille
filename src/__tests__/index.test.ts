import { generateDeck } from '../cards';
import { dealCards, arePartners } from '../index';
import { Player } from '../types';

test('arePartners', () => {
  expect(arePartners(0, 0)).toBe(true);
  expect(arePartners(0, 1)).toBe(false);
  expect(arePartners(0, 2)).toBe(true);
  expect(arePartners(0, 3)).toBe(false);

  expect(arePartners(1, 0)).toBe(false);
  expect(arePartners(1, 1)).toBe(true);
  expect(arePartners(1, 2)).toBe(false);
  expect(arePartners(1, 3)).toBe(true);

  expect(arePartners(2, 0)).toBe(true);
  expect(arePartners(2, 1)).toBe(false);
  expect(arePartners(2, 2)).toBe(true);
  expect(arePartners(2, 3)).toBe(false);

  expect(arePartners(3, 0)).toBe(false);
  expect(arePartners(3, 1)).toBe(true);
  expect(arePartners(3, 2)).toBe(false);
  expect(arePartners(3, 3)).toBe(true);
});

test('dealCards', () => {
  const players: Player[] = [
    { id: 0, cards: [], name: 'a' },
    { id: 1, cards: [], name: 'b' },
    { id: 2, cards: [], name: 'c' },
    { id: 3, cards: [], name: 'd' }
  ];

  dealCards(generateDeck(), players);

  expect(players[0].cards).toHaveLength(8);
  expect(players[1].cards).toHaveLength(8);
  expect(players[2].cards).toHaveLength(8);
  expect(players[3].cards).toHaveLength(8);

  expect(players[0].cards).toMatchSnapshot('player 0');
  expect(players[1].cards).toMatchSnapshot('player 1');
  expect(players[2].cards).toMatchSnapshot('player 2');
  expect(players[3].cards).toMatchSnapshot('player 3');
});
