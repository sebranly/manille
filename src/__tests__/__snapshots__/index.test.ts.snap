// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dealCards: player 0 1`] = `
Array [
  Object {
    "rank": "ten",
    "suit": "clubs",
  },
  Object {
    "rank": "ace",
    "suit": "clubs",
  },
  Object {
    "rank": "king",
    "suit": "clubs",
  },
  Object {
    "rank": "jack",
    "suit": "diamonds",
  },
  Object {
    "rank": "nine",
    "suit": "diamonds",
  },
  Object {
    "rank": "jack",
    "suit": "spades",
  },
  Object {
    "rank": "nine",
    "suit": "spades",
  },
  Object {
    "rank": "eight",
    "suit": "spades",
  },
]
`;

exports[`dealCards: player 1 1`] = `
Array [
  Object {
    "rank": "queen",
    "suit": "clubs",
  },
  Object {
    "rank": "jack",
    "suit": "clubs",
  },
  Object {
    "rank": "nine",
    "suit": "clubs",
  },
  Object {
    "rank": "eight",
    "suit": "diamonds",
  },
  Object {
    "rank": "seven",
    "suit": "diamonds",
  },
  Object {
    "rank": "seven",
    "suit": "spades",
  },
  Object {
    "rank": "ten",
    "suit": "hearts",
  },
  Object {
    "rank": "ace",
    "suit": "hearts",
  },
]
`;

exports[`dealCards: player 2 1`] = `
Array [
  Object {
    "rank": "eight",
    "suit": "clubs",
  },
  Object {
    "rank": "seven",
    "suit": "clubs",
  },
  Object {
    "rank": "ten",
    "suit": "diamonds",
  },
  Object {
    "rank": "ten",
    "suit": "spades",
  },
  Object {
    "rank": "ace",
    "suit": "spades",
  },
  Object {
    "rank": "king",
    "suit": "hearts",
  },
  Object {
    "rank": "queen",
    "suit": "hearts",
  },
  Object {
    "rank": "jack",
    "suit": "hearts",
  },
]
`;

exports[`dealCards: player 3 1`] = `
Array [
  Object {
    "rank": "ace",
    "suit": "diamonds",
  },
  Object {
    "rank": "king",
    "suit": "diamonds",
  },
  Object {
    "rank": "queen",
    "suit": "diamonds",
  },
  Object {
    "rank": "king",
    "suit": "spades",
  },
  Object {
    "rank": "queen",
    "suit": "spades",
  },
  Object {
    "rank": "nine",
    "suit": "hearts",
  },
  Object {
    "rank": "eight",
    "suit": "hearts",
  },
  Object {
    "rank": "seven",
    "suit": "hearts",
  },
]
`;
